### PR TITLE
Refactor cellery build command

### PR DESCRIPTION
**Pull request by MadhukaHarith92**
Refactor cellery build command

### CODE REVIEW

File: components/cli/ballerina/bal_executor.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		return fmt.Errorf("failed to get executable path, %v", err)
+	}
+	cmd = exec.Command(exePath, "run", fileName, "build", string(iName), "{}")
+	executeBuildCommand(cmd)
+	_ = os.Remove(fileName)
+	return nil
+}
+
+// GetBallerinaExecutablePath returns the ballerina executable path.
+func GetBallerinaExecutablePath() (string, error) {
```
Comment (Mirage20) [inline components/cli/ballerina/bal_executor.go]: Do we need to make this public?

File: components/cli/ballerina/bal_executor.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+			execError += stderrScanner.Text()
+		}
+	}()
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err = cmd.Start()
+	if err != nil {
+		errStr := string(stderr.Bytes())
+		fmt.Printf("%s\n", errStr)
```
Comment (Mirage20) [inline components/cli/ballerina/bal_executor.go]: Please remove the print

File: components/cli/ballerina/docker_bal_executor.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	"strings"
+	"time"
+
+	"github.com/cellery-io/sdk/components/cli/pkg/constants"
+	"github.com/cellery-io/sdk/components/cli/pkg/util"
+	"github.com/cellery-io/sdk/components/cli/pkg/version"
+)
+
+type DockerBalExecutor struct {
+	execPath string
```
Comment (Mirage20) [inline components/cli/ballerina/docker_bal_executor.go]: Are there any use of this?

File: components/cli/cli/cli.go [hunk 4]
```diff
@@ -19,27 +19,69 @@
...
 	"os"
+
+	"github.com/cellery-io/sdk/components/cli/ballerina"
+	"github.com/cellery-io/sdk/components/cli/pkg/util"
 )
 
 // Cli represents the cellery command line client.
 type Cli interface {
 	Out() io.Writer
+	ExecuteFunction(startMessage, errorMessage, successMessage string, function func() error) error
```
Comment (Mirage20) [inline components/cli/cli/cli.go]: Shall we rename this to `ExecuteTask`?

File: components/cli/cmd/cellery/cellery.go [hunk 5]
```diff
@@ -83,12 +85,25 @@ func newCliCommand(cli cli.Cli, kubeCli kubernetes.KubeCli) *cobra.Command {
 func main() {
 	celleryCli := cli.NewCelleryCli()
 	celleryKubeCli := kubernetes.NewCelleryKubeCli()
+	var ballerinaExecutor ballerina.BalExecutor
+	moduleMgr := &util.BLangManager{}
+	execPath, err := ballerina.GetBallerinaExecutablePath()
+	if err != nil {
+		util.ExitWithErrorMessage("Failed to get executable path", err)
+	}
```
Comment (Mirage20) [inline components/cli/cmd/cellery/cellery.go]: Can't we move this logic to NewLocalBalExecutor function? and return an error?

File: components/cli/pkg/commands/testdata/hello.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+          global: false
+          authenticate: false
+          destination:
+            host: "hello"
+            port: 80
+        grpc: []
+        tcp: []
+  sts:
+    spec:
+      unsecuredPaths: []
```
Comment (Mirage20) [inline components/cli/pkg/commands/testdata/hello.yaml]: missing new line

File: components/cli/pkg/commands/build.go [hunk 7]
```diff
@@ -19,323 +19,131 @@
...
+	var projectDir string
+	var parsedCellImage *image.CellImage
+	var iName []byte
+	if projectDir, err = cli.FileSystem().CurrentDir(); err != nil {
+		return err
+	}
+	if parsedCellImage, err = image.ParseImageTag(tag); err != nil {
+		return fmt.Errorf("error occurred while parsing image, %v", err)
+	}
+	fmt.Println(parsedCellImage)
```
Comment (Mirage20) [inline components/cli/pkg/commands/build.go]: Please remove print and use cli interface

File: components/cli/ballerina/bal_executor.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+// NewLocalBalExecutor returns a LocalBalExecutor instance.
+func NewLocalBalExecutor(execPath string) *LocalBalExecutor {
+	balExecutor := &LocalBalExecutor{}
+	balExecutor.execPath = execPath
+	return balExecutor
+}
+
+// Build executes ballerina build on an executable bal file.
+func (baleExecutor *LocalBalExecutor) Build(fileName string, iName []byte) error {
+	cmd := &exec.Cmd{}
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Updated.

File: components/cli/ballerina/bal_executor.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	exePath, err := GetBallerinaExecutablePath()
+	if err != nil {
+		return fmt.Errorf("failed to get executable path, %v", err)
+	}
+	cmd = exec.Command(exePath, "run", fileName, "build", string(iName), "{}")
+	executeBuildCommand(cmd)
+	_ = os.Remove(fileName)
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Previous behavior was to remove the temporary executable file after execution.
Moved that logic from this function (Build) as suggested.

File: components/cli/ballerina/bal_executor.go (hunk 2 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Removed.

File: components/cli/pkg/commands/build.go (hunk 7 above)
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/build.go]: Removed.

File: components/cli/cli/cli.go (hunk 4 above)
Comment (MadhukaHarith92) [reply components/cli/cli/cli.go]: Renamed.

File: components/cli/ballerina/bal_executor.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+	return nil
+}
+
+// GetBallerinaExecutablePath returns the ballerina executable path.
+func GetBallerinaExecutablePath() (string, error) {
+	balVersionCmdOutput := ""
+	cmd := exec.Command("ballerina", "version")
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Added a new interface function "Version".

File: components/cli/ballerina/bal_executor.go (hunk 1 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Renamed to "ExecutablePath" and made it an interface function.

File: components/cli/ballerina/docker_bal_executor.go (hunk 3 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: Removed.

File: components/cli/pkg/commands/testdata/hello.yaml (hunk 6 above)
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/testdata/hello.yaml]: Added.

File: components/cli/cmd/cellery/cellery.go (hunk 5 above)
Comment (MadhukaHarith92) [reply components/cli/cmd/cellery/cellery.go]: Created an interface function "ExecutablePath".

File: components/cli/ballerina/bal_executor.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Updated.

File: components/cli/pkg/commands/testdata/metadata.json [hunk 11]
```diff
@@ -0,0 +1 @@
+{"ver":"1.0.0","components":{"hello":{"dockerImage":"wso2cellery/samples-hello-world-webapp","isDockerPushRequired":false,"labels":{},"dependencies":{"composites":{},"components":[],"cells":{}}}},"org":"myorg","autoScalingRequired":false,"kind":"Cell","name":"aaa","zeroScalingRequired":false}
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/testdata/metadata.json]: Added.

File: components/cli/cmd/cellery/build.go [hunk 12]
```diff
@@ -50,7 +52,10 @@ func newBuildCommand() *cobra.Command {
 			return nil
 		},
 		Run: func(cmd *cobra.Command, args []string) {
-			commands.RunBuild(args[1], args[0])
+			err := commands.RunBuild(cli, args[1], args[0])
```
[the remark this answers is not in the document]
Comment (MadhukaHarith92) [reply components/cli/cmd/cellery/build.go]: Updated.

File: components/cli/ballerina/bal_executor.go [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+	var err error
+	var ballerinaVersion string
+	if ballerinaVersion, err = baleExecutor.Version(); err != nil {
+		return "", fmt.Errorf("failed to check ballerina version, %v", err)
+	}
+	if strings.Contains(ballerinaVersion, "Ballerina") {
+		if len(strings.Split(ballerinaVersion, " ")) > 0 {
+			if strings.Split(ballerinaVersion, " ")[1] == constants.BALLERINA_VERSION {
+				// If existing ballerina version is as the expected version, execute ballerina run without executable path
+				return "ballerina", nil
```
Comment (isurulucky) [inline components/cli/ballerina/bal_executor.go]: ballerina is hardcoded in multiple places, move to a constant.

File: components/cli/ballerina/docker_bal_executor.go [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+				if stderrScanner.Scan() && strings.HasPrefix(stderrScanner.Text(), "Unable to find image") {
+					break
+				}
+			}
+		}()
+		go func() {
+			for {
+				if stdoutScanner.Scan() {
+					containerId = []byte(stdoutScanner.Text())
+					break
```
Comment (isurulucky) [inline components/cli/ballerina/docker_bal_executor.go]: do we need a break statement here?

File: components/cli/ballerina/docker_bal_executor.go [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+
+			_, errUserAdd := cmdUserAdd.Output()
+			if errUserAdd != nil {
+				return fmt.Errorf("error in adding Cellery execution user, %v", err)
+			}
+		}
+	}
+	re := regexp.MustCompile("^" + currentDir + "/")
+	balFilePath := re.ReplaceAllString(fileName, "")
+	cmd := exec.Command("docker", "exec", "-w", "/home/cellery/src", "-u", cliUser.Uid,
```
Comment (isurulucky) [inline components/cli/ballerina/docker_bal_executor.go]: move /home/cellery to a const and refer from all places hardcoded.

File: components/cli/cmd/cellery/build.go [hunk 16]
```diff
@@ -21,6 +21,8 @@ package main
 import (
 	"fmt"
 
+	"github.com/cellery-io/sdk/components/cli/cli"
+
```
Comment (isurulucky) [inline components/cli/cmd/cellery/build.go]: format code?

File: components/cli/cmd/cellery/cellery.go [hunk 17]
```diff
@@ -83,12 +85,28 @@ func newCliCommand(cli cli.Cli, kubeCli kubernetes.KubeCli) *cobra.Command {
 func main() {
 	celleryCli := cli.NewCelleryCli()
 	celleryKubeCli := kubernetes.NewCelleryKubeCli()
+	var ballerinaExecutor ballerina.BalExecutor
+	moduleMgr := &util.BLangManager{}
+	// Initially assume ballerina is installed locally and try to get the ballerina executable path.
+	ballerinaExecutor = ballerina.NewLocalBalExecutor()
+	ballerinaExecutablePath, err := ballerinaExecutor.ExecutablePath()
+	if err != nil {
+		util.ExitWithErrorMessage("Failed to get ballerina executable path", err)
```
Comment (isurulucky) [inline components/cli/cmd/cellery/cellery.go]: Return the error.

File: components/cli/internal/test/bal_executor.go [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+	"path/filepath"
+)
+
+type MockBalExecutor struct {
+	CurrentDir string
+}
+
+// Build mocks execution of ballerina build on an executable bal file.
+func (baleExecutor *MockBalExecutor) Build(fileName string, iName []byte) error {
+	os.Create(filepath.Join(baleExecutor.CurrentDir, "metadata.json"))
```
Comment (isurulucky) [inline components/cli/internal/test/bal_executor.go]: handle error.

File: components/cli/internal/test/file_system.go [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+	"io/ioutil"
+)
+
+type mockFileSystem struct {
+}
+
+// CurrentDir returns the current directory.
+func (fs *mockFileSystem) CurrentDir() (string, error) {
+	return ioutil.TempDir("", "cellery-current-dir")
+
```
Comment (isurulucky) [inline components/cli/internal/test/file_system.go]: format code.

File: components/cli/internal/test/file_system.go [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+// UserHome returns the user home.
+func (fs *mockFileSystem) UserHome() string {
+	userHome, _ := ioutil.TempDir("", "user-home")
+	return userHome
```
Comment (isurulucky) [inline components/cli/internal/test/file_system.go]: what happens if an error happens while creating?

File: components/cli/pkg/commands/build.go [hunk 21]
```diff
@@ -19,323 +19,130 @@
...
-		fileCopyError = util.CopyDir(filepath.Join(balTomlParent, constants.BALLERINA_LOCAL_REPO),
-			filepath.Join(projectDir, constants.ZIP_BALLERINA_SOURCE, constants.BALLERINA_LOCAL_REPO))
-	}
-	isTestDirExists, _ := util.FileExists(constants.ZIP_TESTS)
-	folders := []string{constants.ZIP_ARTIFACTS, constants.ZIP_BALLERINA_SOURCE}
-
-	if isTestDirExists {
-		folders = append(folders, constants.ZIP_TESTS)
+	var fileExist bool
+	if fileExist, err = util.FileExists(fileName); err != nil {
```
Comment (isurulucky) [inline components/cli/pkg/commands/build.go]: ideally we should not have this util pkg.. lets handle it later.

File: components/cli/pkg/commands/build_test.go [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+	if diff := cmp.Diff("myorg", got.Organization); diff != "" {
+		t.Errorf("Write (-want, +got)\n%v", diff)
+	}
+	if diff := cmp.Diff("aaa", got.Name); diff != "" {
+		t.Errorf("Write (-want, +got)\n%v", diff)
+	}
+	if diff := cmp.Diff("1.0.0", got.Version); diff != "" {
+		t.Errorf("Write (-want, +got)\n%v", diff)
+	}
+}
```
Comment (isurulucky) [inline components/cli/pkg/commands/build_test.go]: Shall we add a test for RunBuild method.

File: components/cli/ballerina/bal_executor.go (hunk 13 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/bal_executor.go]: Updated.

File: components/cli/ballerina/docker_bal_executor.go (hunk 15 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: Updated.

File: components/cli/cmd/cellery/build.go (hunk 16 above)
Comment (MadhukaHarith92) [reply components/cli/cmd/cellery/build.go]: formatted.

File: components/cli/cmd/cellery/cellery.go (hunk 17 above)
Comment (MadhukaHarith92) [reply components/cli/cmd/cellery/cellery.go]: This is in the main function, hence it does not return to another parent function. Therefore should exit with error message if an error occurs.

File: components/cli/internal/test/bal_executor.go (hunk 18 above)
Comment (MadhukaHarith92) [reply components/cli/internal/test/bal_executor.go]: Updated.

File: components/cli/internal/test/file_system.go (hunk 19 above)
Comment (MadhukaHarith92) [reply components/cli/internal/test/file_system.go]: removed empty line.

File: components/cli/internal/test/file_system.go (hunk 20 above)
Comment (MadhukaHarith92) [reply components/cli/internal/test/file_system.go]: This is to match UserHome method signature of FileSystem interface. However updated UserHome interface function to return string and error.

File: components/cli/pkg/commands/build.go (hunk 21 above)
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/build.go]: Ack.

File: components/cli/pkg/commands/build_test.go (hunk 22 above)
Comment (MadhukaHarith92) [reply components/cli/pkg/commands/build_test.go]: Will add.

File: components/cli/ballerina/docker_bal_executor.go (hunk 14 above)
Comment (MadhukaHarith92) [reply components/cli/ballerina/docker_bal_executor.go]: It seems break statement is not necessary, however this was the previous implementation. Removed it.

**Pull request merged**